### DISCUSSION TITLE
Is ZeRO-3 compatible to tensor parallel?

### DISCUSSION THREAD

**Discussion by skyshine102** [deepspeedai/DeepSpeed#4877] [Q&A] [answered]
In recent blogs posts,

AWS: 
- https://aws.amazon.com/tw/blogs/machine-learning/amazon-sagemaker-model-parallel-library-now-accelerates-pytorch-fsdp-workloads-by-up-to-20/

Nvidia NeMO:
- https://github.com/NVIDIA/NeMo/pull/7897

AWS makes transformer engine's tensor parallel into FSDP, which is "similar" algorithm to ds ZeRO-3. 
  - I'm curious if we can do something similary with ds ZeRO-3? (I know Megatron uses ZeRO-1+TP)
     - If so, how can we wrap sub-modules for TP and the rest for ZeRO-3? While  FSDP provides auto_wrap function, I struggled to find a correspondence in deepspeed.
     
 (Really looking forwards to seeing these hybrid sharding technology being easily accessible)

**Comment by tjruwase** [comment, marked as answer]
@skyshine102, zero-3 can be combined with tensor parallelism and we validated this combination in our [megatron-deespeed](https://github.com/microsoft/Megatron-DeepSpeed/tree/main) repo a while ago. This tutorial might be helpful: https://www.deepspeed.ai/tutorials/zero/

**Reply by skyshine102** [reply]
Cool. Thanks for the pointer. I will take a look
